### PR TITLE
ENH: Make ``__module__`` attribute coherent across API

### PR DESCRIPTION
**Pull request by mtsokol**
Hi!
As described in https://github.com/pytorch/pytorch/issues/136559 `__module__` is incoherent across Python API. I think by definition, function's `__module__` attribute should return the name of the module that this function originates from.

### CODE REVIEW

File: numpy/_core/src/umath/ufunc_object.c [hunk 1]
```diff
@@ -6515,6 +6515,12 @@ ufunc_get_signature(PyUFuncObject *ufunc, void *NPY_UNUSED(ignored))
     return PyUnicode_FromString(ufunc->core_signature);
 }
 
+static PyObject *
+ufunc_get_module(PyUFuncObject *ufunc, void *NPY_UNUSED(ignored))
+{
+    return PyUnicode_FromString("numpy");
```
Comment (mtsokol) [inline numpy/_core/src/umath/ufunc_object.c:6521]: Now I think that it isn't right - `numpy.strings` also have ufuncs, could we add `const char *module` to UFunc struct?

**Comment by rgommers** [comment]
Looks good to me overall so far, thanks for pushing this forward @mtsokol.

> I think by definiton, function's `__module__` attribute should return the name of the module that this function originates from.

To clarify since "originates" could be misunderstood: each function is available in the _public API_ from a single namespace (on rare occasions, some object may be exposed in 2 places still), and we want to ensure that `__module__` points to that public namespace, and not to the private namespace/object where the implementation lives.

File: numpy/_core/src/umath/ufunc_object.c (hunk 1 above)
Comment (mtsokol) [reply numpy/_core/src/umath/ufunc_object.c]: Solved in #27735

**Comment by mtsokol** [comment]
Ok, I think it's ready for reviews. Some comments: 

- I skip all `numpy.random` module (all functions have `__module__` as `numpy.random.mtrand` there instead of `numpy.random`) as, it looks like, cython disallows overriding `__module__` (need to ask about it in the mailing list).
- I also cleaned up `ma` a bit to adhere to "each function is exported from one place" rule, as it was required by my test.

**Comment by mtsokol** [comment]
Update: there are still docs issues to be solved.

**Comment by rgommers** [comment]
> I skip all `numpy.random` module (all functions have `__module__` as `numpy.random.mtrand` there instead of `numpy.random`) as, it looks like, cython disallows overriding `__module__` (need to ask about it in the mailing list).

Ouch, that is bad. In NumPy 1.26 the attribute did contain `numpy.random`, which is more correct. Worth raising upstream indeed.

**Comment by mattip** [comment]
When we discussed this at the triage meeting, we recommended an approach to consider the cython-generated code in a separate OR

**Comment by mattip** [comment]
Thanks @mtsokol. We can adjust this as needed going forward. Hopefully we will find a way to se the cython-generated code in random's `__module__`.

**Pull request merged**